File: source/verify/read_buf_to_file.cc

//read_buf_to_file

#include "rbcp_com.h"
#include "read_buf_to_file.h"
#define BUFSIZE 512

using namespace std;

void read_buf_to_file(const char *ip_addr, unsigned int port, unsigned int buf_addr, unsigned int data_length, const char *filename)
{
   
  char sendData[UDP_BUF_SIZE]; 
  char recvData[UDP_BUF_SIZE];
  char dispMode=0;
  //unsigned int tempInt;

  struct rbcp_header sndHeader;
  sndHeader.type=RBCP_VER;
  sndHeader.command= RBCP_CMD_RD;
  sndHeader.id=0;
  sndHeader.length=data_length;
  sndHeader.address=htonl(buf_addr);

  int rcvdBytes;
  rcvdBytes= rbcp_com(ip_addr, port, &sndHeader, sendData, recvData, dispMode);

  if(rcvdBytes<0){
    cerr << "Receive Error (read_buf_to_file)" << endl;
    exit(EXIT_FAILURE);
  }   

  // if ( (recvData[1] & 0x0F) != 0x08 ) {
  //   cerr << "Receive Error (read_buf_to_file)" << endl;
  //   exit(-1);
  // }

  ofstream file(filename,ios::out|ios::binary|ios::app);

  int datasize=(unsigned char)recvData[3];

  for(int j=8;j<8+datasize;j++)
    file.write(&recvData[j],1);

  file.close();

}

#ifdef TEST_READ_BUF_TO_FILE

int main(int argc,char **argv)
{
  if (argc != 6) {
    cerr << "Usage: "<< argv[0] << " <IP> <port> <rbcp addr> <data length> <output file>" <<endl;
    return 1;
  }
  read_buf_to_file(argv[1], atoi(argv[2]), atoi(argv[3]), atoi(argv[4]), argv[5]);

  return 0;

}

#endif

File: source/verify/read_buf_to_file.py

# read_buf_to_file.py

import sys
import socket
import addr_convert


def read_buf_to_file(ip_addr = '192.168.10.16', port = 4660, buf_addr = 0x800, data_length = 250, filename = "convert_folder/verify_file.txt") :
   
   BUFSIZE = 512
   socket_addr = (ip_addr, port)
   socket.setdefaulttimeout(2.0)
   
   rbcp_ver = chr(255) # 1Byte
   rbcp_cmd = chr(int('0xC0',16))
   rbcp_ID  = chr(0)
   rbcp_len = chr(int(data_length))
   
   int_buf_addr = hex(buf_addr)
   rbcp_addr_list = addr_convert.addr_convert(int_buf_addr)
   
   udpData_list = []
   
   udpData_list.append(rbcp_ver)
   udpData_list.append(rbcp_cmd)
   udpData_list.append(rbcp_ID)
   udpData_list.append(rbcp_len)
   
   udpData_list = udpData_list + rbcp_addr_list
   
   socket_data = "".join(udpData_list)
   
   udp_socket = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
   
   udp_socket.sendto(socket_data,socket_addr)
   socket_data, socket_addr = udp_socket.recvfrom(BUFSIZE)
   
   recvdata_list = list(socket_data)
   
   if ((ord(recvdata_list[1]) & 0x0F) != int('0x08',16)) :
      print "Receive Error"
      sys.exit(-1)
   
   file = open(filename,"ab")
   
   for j in range(8,8+ord(recvdata_list[3])) :
      file.write(recvdata_list[j])
   
   file.close()
   

if __name__ == '__main__' :
   read_buf_cmd(data_length = 250)

